node-wasm/src/ports.rs

```diff
@@ -0,0 +1,283 @@
+use tokio::sync::{mpsc, Mutex};
+use tracing::{error, info, trace};
+use wasm_bindgen::closure::Closure;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::{JsCast, JsValue};
+use web_sys::{MessageEvent, MessagePort};
+
+use crate::commands::{NodeCommand, WorkerResponse};
+use crate::error::{Context, Error, Result};
+use crate::utils::{MessageError, MessageEventExt};
+
+// Instead of supporting communication with just `MessagePort`, allow using any object which
+// provides compatible interface, eg. `Worker`
+#[wasm_bindgen]
+extern "C" {
+    pub type MessagePortLike;
+
+    #[wasm_bindgen(catch, method, structural, js_name = postMessage)]
+    pub fn post_message(this: &MessagePortLike, message: &JsValue) -> Result<(), JsValue>;
+
+    #[wasm_bindgen(catch, method, structural, js_name = postMessage)]
+    pub fn post_message_with_transferable(
+        this: &MessagePortLike,
+        message: &JsValue,
+        transferable: &JsValue,
+    ) -> Result<(), JsValue>;
+}
+
+impl From<MessagePort> for MessagePortLike {
+    fn from(port: MessagePort) -> Self {
+        JsValue::from(port).into()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ClientId(usize);
+
+struct ClientConnection {
+    port: MessagePortLike,
+    _onmessage: Closure<dyn Fn(MessageEvent)>,
+}
+
+impl ClientConnection {
+    fn new(
+        id: ClientId,
+        object: JsValue,
+        forward_messages_to: mpsc::UnboundedSender<(ClientId, Result<NodeCommand, MessageError>)>,
+        forward_connects_to: mpsc::UnboundedSender<JsValue>,
+    ) -> Result<Self> {
+        let onmessage = Closure::new(move |ev: MessageEvent| {
+            let message = ev.get_message();
+
+            if let Some(port) = ev.get_port() {
+                if let Err(e) = forward_connects_to.send(port) {
+                    error!("port forwarding channel closed, shouldn't happen: {e}");
+                }
+            }
+
+            if let Err(e) = forward_messages_to.send((id, message)) {
+                error!("message forwarding channel closed, shouldn't happen: {e}");
+            }
+        });
+
+        let port = prepare_message_port(object, &onmessage)
+            .context("failed to setup port for ClientConnection")?;
+
+        Ok(ClientConnection {
+            port,
+            _onmessage: onmessage,
+        })
+    }
+
+    fn send(&self, message: &WorkerResponse) -> Result<()> {
+        let serializer = Serializer::json_compatible();
+        let message_value = message
+            .serialize(&serializer)
+            .context("could not serialise message")?;
+        self.port
+            .post_message(&message_value)
+            .context("could not send command to worker")?;
+        Ok(())
+    }
+}
+
+pub struct WorkerServer {
+    ports: Vec<ClientConnection>,
+    connect_tx: mpsc::UnboundedSender<JsValue>,
+    connect_rx: mpsc::UnboundedReceiver<JsValue>,
+    request_tx: mpsc::UnboundedSender<(ClientId, Result<NodeCommand, MessageError>)>,
```

Comment:
I thought that after the signature change of the stuff, we could simplify this to `mpsc::UnboundedSender<(ClientId, NodeCommand)>`. would that be possible? Could we just log the error at onmessage level and not pass it further?

Reply:
I think we can just ignore the un-deserialisable messages in the callback for the server side :+1: 
I've left client side as is, with how `exec` is implemented, it waits for some kind of response error or not.